### ISSUE TITLE
ts_project ignores TypeScript errors when supports_workers = True

### ISSUE THREAD

**Issue by gregjacobs** [bazelbuild/rules_nodejs#3440]
# 🐞 bug report

### Affected Rule

 ts_project

### Is this a regression?

Not sure

### Description

When using `ts_project` with `supports_workers = True`, TypeScript errors are reported to the console, but the build incorrectly succeeds.

When `supports_workers = False` (the default), the build properly fails when there are TypeScript errors.

## 🔬 Minimal Reproduction

https://github.com/gregjacobs/rules-nodejs-supports-workers-issue

## 🔥 Exception or Error

The build succeeds, even though there is a TypeScript error:
<pre><code>$ bazel build //package-1:tsc
INFO: Analyzed target //package-1:tsc (1 packages loaded, 5 targets configured).
INFO: Found 1 target...
INFO: From Compiling TypeScript project (worker mode) //package-1:tsc [tsc -p package-1/tsconfig.json]:
package-1/index.ts(1,7): error TS2322: Type 'number' is not assignable to type 'string'.

Target //package-1:tsc up-to-date:
  bazel-bin/package-1/index.js
  bazel-bin/package-1/index.js.map
INFO: Elapsed time: 3.699s, Critical Path: 3.47s
INFO: 41 processes: 15 internal, 1 darwin-sandbox, 24 local, 1 worker.
INFO: Build completed successfully, 41 total actions
</code></pre>

## 🌍  Your Environment

**Operating System:** 

<pre><code>Mac OSX 10.14.6</code></pre>

**Output of `bazel version`:**

<pre><code>Bazelisk version: v1.11.0
Build label: 5.1.1
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Apr 8 15:57:36 2022 (1649433456)
Build timestamp: 1649433456
Build timestamp as int: 1649433456</code></pre>

**Rules_nodejs version:**

<pre><code>5.4.2</code></pre>

**Comment by alexeagle** [comment]
oh noes.
easy to repro at HEAD with a broken edit to `packages/typescript/test/ts_project/worker/big.ts` as well

**Comment by alexeagle** [comment]
turned on DEBUG in packages/worker/src/worker.ts
and added logging
`console.error(result)`
at https://github.com/bazelbuild/rules_nodejs/blob/stable/packages/typescript/internal/worker/worker_adapter.ts#L114
and `result` from `watchProgram ?.getProgram().emit` is printed at the end

```
INFO: From Compiling TypeScript project (worker mode) //packages/typescript/test/ts_project/worker:tsconfig [tsc -p packages/typescript/test/ts_project/worker/tsconfig_tsconfig.json]:
Reiterating
We do not have the full message yet. Keep reading
Reiterating
Now we have the full message. Time to go!
Handling new build request: undefined
Compiling with:
        --project
        bazel-out/darwin_arm64-fastbuild/bin/packages/typescript/test/ts_project/worker/tsconfig_tsconfig.json
        --outDir
        bazel-out/darwin_arm64-fastbuild/bin/packages/typescript/test/ts_project/worker
        --rootDir
        packages/typescript/test/ts_project/worker
        --declarationDir
        bazel-out/darwin_arm64-fastbuild/bin/packages/typescript/test/ts_project/worker
message TS6031: Starting compilation in watch mode...

packages/typescript/test/ts_project/worker/big.ts(3,14): error TS2322: Type 'string' is not assignable to type 'number'.

message TS6193: Found 1 error. Watching for file changes.

{
  emitSkipped: false,
  diagnostics: [],
  emittedFiles: [],
  sourceMaps: []
}
Compilation complete for request_id undefined. Success: true
```

@mrmeku or @thesayyn you were in this code most recently, any ideas?

**Comment by alexeagle** [comment]
maybe the `reportDiagnostic` function we pass to `createWatchCompilerHost` causes the diagnostics to go to the worker.log but not end up in the emit `result`? I don't know this API well enough to guess.

**Comment by thesayyn** [comment]
At first sight, it looks like a problem in the typescript package rather than the worker package. 
 
`Compilation complete for request_id undefined. Success: true`  this line gives us the clue

**Comment by thesayyn** [comment]
> maybe the `reportDiagnostic` function we pass to `createWatchCompilerHost` causes the diagnostics to go to the worker.log but not end up in the emit `result`? I don't know this API well enough to guess.

`emitResult.diagnostics` reports the diagnostics from transformers and other emitter internals hence when you have a typechecking diagnostic it is not going to appear in `emitResult.diagnostics`

**Comment by alexeagle** [comment]
will be fixed in the next release. Do you think we need to back-port to 4.x as well?

**Comment by gregjacobs** [comment]
Thanks for the quick turnaround on this @alexeagle and @thesayyn!

**Comment by thesayyn** [comment]
> will be fixed in the next release. Do you think we need to back-port to 4.x as well?

Yeah. as the users are still a bit hesitant to upgrade to 5.x.  Will make a follow-up PR.
